### PR TITLE
feat: Integrate moralis as nft provider

### PR DESCRIPTION
**Pull request by AndonMitev**

### CODE REVIEW

File: packages/evm/lib/nft/MoralisNftProvider.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { ClientTypes } from '@chainify/client';
+import { BaseProvider } from '@ethersproject/providers';
+import { EvmNftProvider } from 'lib/nft/EvmNftProvider';
```
Comment (kraikov) [inline packages/evm/lib/nft/MoralisNftProvider.ts]: fix this path

File: packages/evm/lib/nft/OpenSeaNftProvider.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { ClientTypes } from '@chainify/client';
+import { BaseProvider } from '@ethersproject/providers';
+import { NFTAsset, NftTypes } from 'lib/types';
```
Comment (kraikov) [inline packages/evm/lib/nft/OpenSeaNftProvider.ts]: path

**Review by kraikov — CHANGES_REQUESTED** [review]

File: packages/evm/lib/nft/OpenSeaNftProvider.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+export class OpenSeaNftProvider extends EvmNftProvider {
+    constructor(walletProvider: EvmBaseWalletProvider<BaseProvider>, httpConfig: ClientTypes.AxiosRequestConfig) {
+        super(walletProvider, httpConfig);
+    }
+
+    async fetch(): Promise<NFTAsset[]> {
+        const userAddress = await this.walletProvider.getAddress();
+        const nfts = await this.httpClient.nodeGet(`assets?owner=${userAddress}`);
+
+        const data = nfts.assets.map((nft: any) => {
```
Comment (monokh) [inline packages/evm/lib/nft/OpenSeaNftProvider.ts:16]: no :any, type the response from opensea

Comment (monokh) [inline packages/evm/lib/nft/OpenSeaNftProvider.ts]: no :any, type the response from opensea

File: packages/evm/lib/nft/MoralisNftProvider.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+export class MoralisNftProvider extends EvmNftProvider {
+    constructor(
+        walletProvider: EvmBaseWalletProvider<BaseProvider>,
+        httpConfig: ClientTypes.AxiosRequestConfig,
+        moralisConfig: MoralisConfig
+    ) {
+        super(walletProvider, httpConfig);
+
+        Moralis.start(moralisConfig).then();
```
Comment (monokh) [inline packages/evm/lib/nft/MoralisNftProvider.ts]: what is going on here? if `.start()` needs to be awaited, you can't do it like this. You can call `start` on demand when `fetch` is called for example. And after it's started, not call it again

File: packages/evm/lib/nft/MoralisNftProvider.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+            return {
+                ...nftAsset,
+                name: parsed.name,
+                description: parsed.description,
+                image_original_url: parsed.image,
+                image_preview_url: parsed.image,
+                image_thumbnail_url: parsed.image,
+                external_link: parsed.external_url,
+            } as NFTAsset;
```
Comment (monokh) [inline packages/evm/lib/nft/MoralisNftProvider.ts]: `as NFTAsset` is usually a sign that the typing here is not correct and you are force ignoring some type errors. What happens if you remove it ?

File: packages/evm/lib/nft/CovalentNftProvider.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                        },
+                    };
+
+                    this.cache[contract_address] = {
+                        contract: this.schemas[supports_erc.pop().toUpperCase()].attach(contract_address),
+                        schema: supports_erc.pop().toUpperCase() as NftTypes,
+                    };
+
+                    if (!nft_data.length) {
+                        return resp as NFTAsset;
```
Comment (monokh) [inline packages/evm/lib/nft/CovalentNftProvider.ts]: shouldn't need `as` if you define the response type of the api

File: packages/evm/lib/nft/EvmNftProvider.ts [hunk 7]
```diff
@@ -1,32 +1,34 @@
...
     private _erc1155: ERC1155;
-    private _cache: Record<string, NftInfo>;
-    private _schemas: Record<string, NftContract>;
-    private _httpClient: HttpClient;
+
+    protected schemas: Record<string, NftContract>;
+    protected cache: Record<string, NftInfo>;
+    protected httpClient: HttpClient;
 
     constructor(walletProvider: EvmBaseWalletProvider<BaseProvider>, httpConfig: ClientTypes.AxiosRequestConfig) {
```
Comment (monokh) [inline packages/evm/lib/nft/EvmNftProvider.ts]: replace `httpConfig` with the parameters that client needs to pass. e.g. `url`, `apiKey`

File: packages/evm/lib/nft/CovalentNftProvider.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                        name: external_data.name,
+                        description: external_data.description,
+                        external_link: external_data.external_url,
+                        image_original_url: external_data.image,
+                        image_preview_url: external_data.image,
+                        image_thumbnail_url: external_data.image,
+                    } as NFTAsset;
+                }
+            })
+            .filter(Boolean);
```
Comment (monokh) [inline packages/evm/lib/nft/CovalentNftProvider.ts]: I think you are looking for `reduce` 😄  Inside reduce you can check if nft, and if so convert and add to the accumulator

File: packages/evm/package.json [hunk 9]
```diff
@@ -30,7 +30,8 @@
         "@ethersproject/transactions": "5.6.0",
         "@ethersproject/wallet": "5.6.0",
         "@metamask/eth-sig-util": "^4.0.1",
-        "@rainbow-me/fee-suggestions": "^2.0.0"
+        "@rainbow-me/fee-suggestions": "^2.0.0",
+        "moralis": "^1.8.0"
```
Comment (monokh) [inline packages/evm/package.json]: pin the version

File: packages/evm/lib/nft/MoralisNftProvider.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                asset_contract: {
+                    address: token_address,
+                    name,
+                    symbol,
+                },
+                collection: {
+                    name,
+                },
+                token_id,
+            } as NFTAsset;
```
Comment (monokh) [inline packages/evm/lib/nft/MoralisNftProvider.ts]: same as below

**Review by monokh — CHANGES_REQUESTED** [review]

**Comment by kraikov** [comment]
Usage:

```ts
import * as EVM from "@chainify/evm";

const infuraApi = "";
const mnemonic = "";

const chainProvider = new EVM.EvmChainProvider(
  {
    ...EVM.EvmNetworks.rinkeby,
    rpcUrl: `https://rinkeby.infura.io/v3/${infuraApi}`,
  },
  null,
  null,
  false
);

const walletProvider = new EVM.EvmWalletProvider({ mnemonic }, chainProvider);

// Fetch NFTs from OpenSea
const openSea = new EVM.OpenSeaNftProvider(walletProvider, {
  url: "https://rinkeby-api.opensea.io/api/v1/",
  apiKey: undefined, // define for mainnet
});
openSea.fetch().then(openSea.fetch().then(console.log()));

// Fetch NFTs from Moralis
const moralis = new EVM.MoralisNftProvider(walletProvider, {
  url: "",
  apiKey: "",
  appId: "",
});
moralis.fetch().then(console.log);

// Fetch NFTs from Covalent (does not work with Rinkeby)
const covalent = new EVM.CovalentNftProvider(walletProvider, {
  url: "https://api.covalenthq.com/v1",
  apiKey: "",
});
covalent.fetch().then(console.log);

// Transfer
moralis
  .transfer(
    "0xe10cd16070101ffe6b5bdd380c3c7f0e8fcec84d", // nft contract address
    "0x8072aC361d0ee237EFE0883198613d8d121Ee98d", // recipient
    ["1"], // token ids
    ["1"] // token amounts only applicable for ERC1155
  )
  .then(console.log);

```

**Comment by kraikov** [comment]
Responses:

OpenSea:

```js
[
  {
    asset_contract: {
      address: '0xe10cd16070101ffe6b5bdd380c3c7f0e8fcec84d',
      external_link: null,
      image_url: null,
      name: 'test1',
      symbol: 'test1'
    },
    collection: { name: 'test1 - uRJXEljW05' },
    description: 'Bitcoin, often described as a cryptocurrency, a virtual currency or a digital currency - is a type of money that is completely virtual.',
    external_link: null,
    id: 57292149,
    image_original_url: 'ipfs://QmWsDVqrE5wp2ct6CxXzRaZeG8QxuJR7ALiuLaKrJ1UvYf/0000000000000000000000000000000000000000000000000000000000000001.png',
    image_preview_url: 'https://lh3.googleusercontent.com/rfjtgTE04GAX2WhKqDTFx1ZHd-uJEdhqD9XCyIDurHTr5KoBthbAl1s_LUJ1-2XSVZRoObd40J68mNuq8QpX7ZNBdoWfdr-JCA7fU5c=s250',
    image_thumbnail_url: 'https://lh3.googleusercontent.com/rfjtgTE04GAX2WhKqDTFx1ZHd-uJEdhqD9XCyIDurHTr5KoBthbAl1s_LUJ1-2XSVZRoObd40J68mNuq8QpX7ZNBdoWfdr-JCA7fU5c=s128',
    name: 'test1',
    token_id: '1'
  }
]
```

Moralis:
```js
[
  {
    asset_contract: {
      address: '0xe10cd16070101ffe6b5bdd380c3c7f0e8fcec84d',
      name: 'test1',
      symbol: 'test1'
    },
    collection: { name: 'test1' },
    token_id: '1',
    name: 'Bitcoin 1',
    description: 'Bitcoin, often described as a cryptocurrency, a virtual currency or a digital currency - is a type of money that is completely virtual.',
    image_original_url: 'ipfs://QmWsDVqrE5wp2ct6CxXzRaZeG8QxuJR7ALiuLaKrJ1UvYf/0000000000000000000000000000000000000000000000000000000000000001.png',
    image_preview_url: 'ipfs://QmWsDVqrE5wp2ct6CxXzRaZeG8QxuJR7ALiuLaKrJ1UvYf/0000000000000000000000000000000000000000000000000000000000000001.png',
    image_thumbnail_url: 'ipfs://QmWsDVqrE5wp2ct6CxXzRaZeG8QxuJR7ALiuLaKrJ1UvYf/0000000000000000000000000000000000000000000000000000000000000001.png',
    external_link: undefined
  }
]
```


Covalent
``` js
[
  {
    token_id: '36500',
    asset_contract: {
      address: '0xaaf03a65cbd8f01b512cd8d530a675b3963de255',
      name: 'Seekers',
      symbol: 'SEEKERS'
    },
    collection: { name: 'Seekers' },
    name: null,
    description: null,
    external_link: 'https://seekers.xyz',
    image_original_url: 'https://nft.seekers.xyz/seekers/seeker_36500.png',
    image_preview_url: 'https://nft.seekers.xyz/seekers/seeker_36500.png',
    image_thumbnail_url: 'https://nft.seekers.xyz/seekers/seeker_36500.png'
  }
]
```

**Review by monokh — APPROVED** [review]

**Pull request merged**